File: Data Structures & Algorithms/depth-of-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxDepth(TreeNode* root) {
        int depth = 0;

        queue<TreeNode*> q;

        if (root != nullptr)
            q.push(root);

        while(!q.empty()) {

            depth++;
            int currentSize = q.size();

            for(int i = 0; i < currentSize; i++) {

                // Empty the queue between each iteration
                TreeNode* qt = q.front();
                q.pop();

                if (qt->left)
                    q.push(qt->left);
            
                if (qt->right)
                    q.push(qt->right);
            }

        }

        return depth;
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        self.maxD = 0

        def dfs(n: Optional[TreeNode]):
            if not n:
                return 0

            left_h = dfs(n.left)
            right_h = dfs(n.right)

            self.maxD = max(self.maxD, left_h + right_h)
            
            return max(left_h, right_h) + 1

        dfs(root)

        return self.maxD